File: C++/Array/repeating_missing.cpp
/*
 * DATE: April 5, 2026
 * PATTERN: Hashing / Counting
 *
 * APPROACHES:
 * 1. BRUTE FORCE:
 * - Step 1: For each number 1..n, count occurrences by scanning the whole array each time.
 * - TIME: O(n^2) | SPACE: O(1)
 *
 * 2. BETTER:
 * - Step 1: Frequency map or array of size n+1; increment per value; find value with count 2 and missing 0.
 * - TIME: O(n) | SPACE: O(n)
 *
 * 3. OPTIMAL (math variants exist):
 * - Step 1: Use counting array as in “better” when values in 1..n—same asymptotics, minimal overhead.
 * - Alternative: sum and sum of squares equations to derive repeating and missing (O(n) time, O(1) space).
 * - TIME: O(n) | SPACE: O(n) for frequency array here, or O(1) with algebraic method
 *
 * AHA-MOMENT: One duplicate and one missing show up as count 2 and count 0 in a 1..n index frequency table.
 */

#include<bits/stdc++.h>
using namespace std;

class Solution
{
    public:
    vector<int> repeating_missing(vector<int> &arr)
    {
        int n=arr.size();
        vector<int> hash(n+1,0);
        vector<int>ans;
        for(int i=0;i<n;i++)n
        {
            hash[arr[i]]++;
        }
        for(int i=0;i<hash.size();i++)
        {
            if(hash[i]==2)
            {
                ans.push_back(i);
            }
            if(hash[i]==0)
            {
                ans.push_back(i);
            }
        }
        return ans;
    }
};

int main() {
    vector<int> nums = {3, 1, 2, 5, 4, 6, 7, 5};
    
    // Create an instance of Solution class
    Solution sol;

    vector<int> result = sol.repeating_missing(nums);
    
    // Print the repeating and missing numbers found
    cout << "The repeating and missing numbers are: {" << result[0] << ", " << result[1] << "}\n";
    
    return 0;
}

File: Python/Array/repeating_missing.py
"""
DATE: April 5, 2026
PATTERN: Math (sum and sum of squares), Hashing

APPROACHES:
1. BRUTE FORCE:
- Step 1: Count frequency of each value by scanning array repeatedly.
- TIME: O(n^2) | SPACE: O(1)

2. BETTER:
- Step 1: Hash map frequencies; find duplicate count 2 and missing in 1..n.
- TIME: O(n) | SPACE: O(n)

3. OPTIMAL:
- Step 1: Use expected sum and sum of squares vs actual to solve for the two unknown values (two equations).
- TIME: O(n) | SPACE: O(1)

AHA-MOMENT: Sum and sum-of-squares give two linear equations in the repeating and missing numbers.
"""

from collections import defaultdict


class Solution:
    # def repeat_missing(self,arr):
    #     freq={}
    #     repeating=-1
    #     missing=-1
        
    #     for i in range(len(arr)):
    #         freq[arr[i]]=freq.get(arr[i],0)+1
    #         if freq[arr[i]]==2:
    #             repeating=arr[i]

    #     for i in range(1,len(arr)+1):
    #         if i not in freq:
    #             missing=i
    #             break

    #     return repeating,missing


    def repeat_missing(self, arr):
        n=len(arr)
        s,sn=0,0
        s2=(n*(n+1))//2
        s2n=(n*(n+1)*(2*n+1))//6

        for i in range(n):
            s+=arr[i]
            sn+=(arr[i]**2)

        val1=s2-s
        val2=s2n-sn

        val2=val2//val1

        x=(val1+val2)//2
        y=x-val1

        return x,y


if __name__=="__main__":
    arr=[3, 1, 2, 5, 4, 6, 7, 5]
    sol=Solution()
    a,b=sol.repeat_missing(arr)
    print(f"The repeating number is {a} and the missing number is {b}")
